README.md


# @Al00X/Screen-Detector
> 🖨️ _Get your device screen type reactively!_ in Angular.


## ✅ Setup:
Install using your preferred package manager:

    npm i @al00x/screen-detector

## ✨ Usage:
`ScreenDetectorService` it's what you need!

```html
<div *ngIf="screenDetector.isDesktop$ | async">
  Render when reached desktop breakpoint
</div>

<!-- The below *ngIf condition, it's always true -->
<ng-container *ngIf="{isXL: screenDetector.xl$ | async} as device">
  XL Breakpoint? {{ device.isXL ? 'YES' : 'NO'}}
</ng-container>
```
There's 5 different breakpoint types, also with utility: `isDesktop` and `state`.

### ⚙️ Config:
You can config screen detector for more customization by providing `ALX_SCREEN_DETECTOR_CONFIG` in **AppModule**.

```ts
// app.module.ts

@NgModule({
  ...,
  providers: [
    {
      provide: ALX_SCREEN_DETECTOR_CONFIG,
      useValue: {
        ...,
        // The breakpoints below are the default values that are the same as default TailwindCSS breakpoints.
        breakpoints: {
          xxl: 1536,
          xl: 1280,
          lg: 1024,
          md: 768,
          sm: 640,
        }
      } as AlxScreenDetectorConfig
    }
  ],
})
export class AppModule { }
```

#### Config Options:

| property | default | description |
|--|--|--|
| `desktopBreakpoint: 'xxl' \| 'xl' \| 'lg' \| 'md' \| 'sm'`  | 'lg' | You can set at which breakpoint, the screen is considered as desktop |
| `resizeDebounceTime: number`  | 25 | Screen resize event debounce time in milliseconds **(affects performance)** |
| `breakpoints: BreakpointsConfig`  | *In the example above* | Describe each breakpoint's pixels |




CHANGELOG.md
## 1.1.0
Added `state$` utility property which contains all the breakpoints state
## 1.0.3
Documentation updated
## 1.0.2
Documentation updated
## 1.0.1
Documentation updated
## 1.0.0
Initial release


playground/src/app/app.component.spec.ts
import { TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { AppComponent } from './app.component';

describe('AppComponent', () => {
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        RouterTestingModule
      ],
      declarations: [
        AppComponent
      ],
    }).compileComponents();
  });

  it('should create the app', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    expect(app).toBeTruthy();
  });

  it(`should have as title 'playground'`, () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.componentInstance;
    expect(app.title).toEqual('playground');
  });

  it('should render title', () => {
    const fixture = TestBed.createComponent(AppComponent);
    fixture.detectChanges();
    const compiled = fixture.nativeElement as HTMLElement;
    expect(compiled.querySelector('.content span')?.textContent).toContain('playground app is running!');
  });
});


playground/karma.conf.js
// Karma configuration file, see link for more information
// https://karma-runner.github.io/1.0/config/configuration-file.html

module.exports = function (config) {
  config.set({
    basePath: '../projects/playground',
    frameworks: ['jasmine', '@angular-devkit/build-angular'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage'),
      require('@angular-devkit/build-angular/plugins/karma')
    ],
    client: {
      jasmine: {
        // you can add configuration options for Jasmine here
        // the possible options are listed at https://jasmine.github.io/api/edge/Configuration.html
        // for example, you can disable the random execution with `random: false`
        // or set a specific seed with `seed: 4321`
      },
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },
    jasmineHtmlReporter: {
      suppressAll: true // removes the duplicated traces
    },
    coverageReporter: {
      dir: require('path').join(__dirname, '../../coverage/playground'),
      subdir: '.',
      reporters: [
        { type: 'html' },
        { type: 'text-summary' }
      ]
    },
    reporters: ['progress', 'kjhtml'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: false,
    restartOnFileChange: true
  });
};


src/lib/screen-detector.service.ts
import {Inject, Injectable, InjectionToken, OnDestroy, Optional} from '@angular/core';
import {BehaviorSubject, debounceTime, fromEvent, Subscription} from 'rxjs';

type BreakpointKeys = 'xxl' | 'xl' | 'lg' | 'md' | 'sm';
export type BreakpointsConfig = {[key in BreakpointKeys]: number};
export interface AlxScreenDetectorConfig {
  resizeDebounceTime?: number;
  desktopBreakpoint?: BreakpointKeys;
  breakpoints?: BreakpointsConfig;
}

export const ALX_SCREEN_DETECTOR_CONFIG = new InjectionToken('Config for AlX Screen Detector package');

/* The default breakpoints are the same as TailwindCSS sizes */
const DEFAULT_BREAKPOINTS: BreakpointsConfig = {
  xxl: 1536,
  xl: 1280,
  lg: 1024,
  md: 768,
  sm: 640,
}
const DEFAULT_DESKTOP_BREAKPOINT = 'lg';
const DEFAULT_RESIZE_DEBOUNCE_TIME = 25;

@Injectable({
  providedIn: 'root',
})
export class ScreenDetectorService implements OnDestroy {
  public xxl$ = new BehaviorSubject<boolean>(false);
  public xl$ = new BehaviorSubject<boolean>(false);
  public lg$ = new BehaviorSubject<boolean>(false);
  public md$ = new BehaviorSubject<boolean>(false);
  public sm$ = new BehaviorSubject<boolean>(false);
  public isDesktop$ = new BehaviorSubject<boolean>(false);
  public state$ = new BehaviorSubject<{[key in BreakpointKeys | 'isDesktop']: boolean}>({
    xxl: false,
    xl: false,
    lg: false,
    md: false,
    sm: false,
    isDesktop: false,
  })

  private resizeSub$: Subscription;
  private readonly breakpoints: BreakpointsConfig;
  private readonly desktopBreakpointKey: BreakpointKeys;
  private readonly resizeDebounceTime: number;

  constructor(@Optional() @Inject(ALX_SCREEN_DETECTOR_CONFIG) private config?: AlxScreenDetectorConfig) {
    this.breakpoints = config?.breakpoints ?? DEFAULT_BREAKPOINTS;
    this.desktopBreakpointKey = config?.desktopBreakpoint ?? DEFAULT_DESKTOP_BREAKPOINT;
    this.resizeDebounceTime = config?.resizeDebounceTime ?? DEFAULT_RESIZE_DEBOUNCE_TIME;

    this.resizeSub$ = fromEvent(window, 'resize').pipe(debounceTime(this.resizeDebounceTime)).subscribe(() => {
      this.updateState();
    });
    this.updateState();
  }

  private updateState() {
    const screenWidth = window.innerWidth;
    this.sm$.next(screenWidth >= this.breakpoints.sm);
    this.md$.next(screenWidth >= this.breakpoints.md);
    this.lg$.next(screenWidth >= this.breakpoints.lg);
    this.xl$.next(screenWidth >= this.breakpoints.xl);
    this.xxl$.next(screenWidth >= this.breakpoints.xxl);

    this.isDesktop$.next(screenWidth >= this.breakpoints[this.desktopBreakpointKey]);

    this.state$.next({
      sm: this.sm$.value,
      md: this.md$.value,
      lg: this.lg$.value,
      xl: this.xl$.value,
      xxl: this.xxl$.value,
      isDesktop: this.isDesktop$.value,
    })
  }

  ngOnDestroy() {
    this.resizeSub$.unsubscribe();
  }
}
